Concurrent copy

sargun opened the pull request:
**- What I did**
I made copies concurrent for VFS, and Overlay1. 

**- How I did it**
I started with adding a benchmark to the tests to give me a baseline. On my test system that gave me about ~1081927574 ns/op. I slowly improved the performance throughout the commits, and increased performance to about 552743671 ns/op in the final tests. I parallizes 

**- How to verify it**
There are attached benchmarks, and tests

**- Description for the changelog**
Make VFS Faster.

Commit:
copy: Add benchmark, and size checking to tests

This adds a basic benchmark function, and writes
some data during the tests for copy.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

Commit:
copy: Refactor, and separate out metadata copy

This is a step to begin copying the file data concurrently
with the metadata. It separates out the metadata functions
which are used for standard file copying, symlink copies,
and directory copies.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

Commit:
copy: Copy file data in the background, while copying metadata

This is based on the following benchmarks on EXT4 filesystem
with the built-in benchmark test:
No background copy:
1081927574 ns/op
1082520482 ns/op

Background copy, chan size 1:
1569585224 ns/op

Background copy, chan size 10:
924814623 ns/op

Background copy, chan size 100:
643039679 ns/op

Background copy, chan size 500:
698984799 ns/op

It plateaued at ~100, and then started getting slower as I went
further.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

Commit:
copy: Process directory metadata concurrently

This also moves copying of directory metadata into the background
copying workers. This shows about another 15% in terms of speed
on EXT4.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

sargun commented:
FWIW, benchmark results (of the included synthetic benchmark):

```
XFS:
After improvements:
BenchmarkCopyDir-64    	      30	 804674003 ns/op
Before improvements:
BenchmarkCopyDir-64    	      20	1477743492 ns/op


EXT4
After improvements:
BenchmarkCopyDir-64    	      50	 578406619 ns/op
Before improvements:
BenchmarkCopyDir-64    	      20	1206396460 ns/op
```

This is a pretty significant performance improvement for most folks who want to use VFS.

sargun commented:
For context of why we're suddenly so interested in this: https://www.spinics.net/lists/linux-unionfs/msg05779.html -- overlay2 is a pretty big pain in production.

Commit:
copy: Add small optimization to remove unncessary chmod call

All syscalls we make to make files set the permissions bits. If
are setting any permissions bits that conflict with the umask,
then we need to call chmod, otherwise our initial file creation
call will work fine.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

Commit:
copy: Do dumb multiplexing of copiers, so we have parallel file copy

This starts 8 workers (tuned) based on EXT4, and XFS results, which
inodes are randomly sorted across and data is copied.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "concurrent-copy" git@github.com:sargun/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354000632
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

sargun commented:
Some more performance numbers:

```
sargun@ubuntu:~/go/src/github.com/docker/docker$ time docker run ubuntu:latest echo


real	0m0.576s
user	0m0.056s
sys	0m0.017s
sargun@ubuntu:~/go/src/github.com/docker/docker$ docker info -f {{.Driver}}
vfs
sargun@ubuntu:~/go/src/github.com/docker/docker$ time docker run ubuntu:latest echo


real	0m0.906s
user	0m0.057s
sys	0m0.015s
sargun@ubuntu:~/go/src/github.com/docker/docker$ docker image inspect ubuntu --format {{.Size}}
84114788
sargun@ubuntu:~/go/src/github.com/docker/docker$ docker image inspect ubuntu |jq '.[0].RootFS.Layers | length'
5

```

sargun commented:
@dmcgowan If you want, we can split the patchset up. It comes in pieces nicely.

Commit:
copy: Copy directory mtimes asyncronously

This has quite a few changes in it as it brings in the filepath
walkdir code into copy, so we can make further speed enhancements
here.

It now copies the mtime of the directory using the same mechanism
as the other file data / metadata copies, and it serializes writes
to the diven directory in order to make sure that's the last "update"
that arrives.

Lastly, it fixes problems with failures, and deadlocks around blocked
channels if the workers exit, but our walk continues.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

sargun commented:
So, just to get some benchmarks, (The ones with TMPDIR are on XFS, the other on EXT4):
```
TMPDIR=/mnt/bench ./copy.be29f743ef -test.run none  -test.bench=.* -test.benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	   20000	    346803 ns/op
BenchmarkCopyDir-64                      	      10	 874673948 ns/op
PASS
./copy.be29f743ef -test.run none  -test.bench=.* -test.benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	   20000	    269299 ns/op
BenchmarkCopyDir-64                      	      10	 702427282 ns/op
PASS


158a50c4ff (HEAD) copy: Add small optimization to remove unncessary chmod call
TMPDIR=/mnt/bench ./copy.158a50c4ff  -test.run none  -test.bench=.* -test.benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	   20000	    440871 ns/op
BenchmarkCopyDir-64                      	      10	1014970563 ns/op
PASS
$ ./copy.158a50c4ff  -test.run none  -test.bench=.* -test.benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	   30000	    321325 ns/op
BenchmarkCopyDir-64                      	      10	 831401364 ns/op
PASS


24050449be (HEAD) copy: Process directory metadata concurrently
$ TMPDIR=/mnt/bench ./copy.24050449be  -test.run none  -test.bench=.* -test.benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	   20000	    467087 ns/op
BenchmarkCopyDir-64                      	      10	1025676831 ns/op
PASS
$ ./copy.24050449be  -test.run none  -test.bench=.* -test.benchtime 2s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	   10000	    230710 ns/op
BenchmarkCopyDir-64                      	     100	 440939286 ns/op
PASS


e6ffc8a646 (HEAD) copy: Do dumb multiplexing of copiers, so we have parallel file copy
$ TMPDIR=/mnt/bench ./copy.e6ffc8a646  -test.run none  -test.bench=.* -test.benchtime 2s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	     500	   5114880 ns/op
BenchmarkCopyDir-64                      	       2	1165530186 ns/op
PASS
$ ./copy.e6ffc8a646  -test.run none  -test.bench=.* -test.benchtime 2s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	    1000	   2886859 ns/op
BenchmarkCopyDir-64                      	       3	2650059962 ns/op
PASS

6e48e3fb21 (HEAD) copy: Copy file data in the background, while copying metadata
$ TMPDIR=/mnt/bench ./copy.6e48e3fb21  -test.run none  -test.bench=.* -test.benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	   20000	    300600 ns/op
BenchmarkCopyDir-64                      	      10	1097935740 ns/op
PASS
$ ./copy.6e48e3fb21  -test.run none  -test.bench=.* -test.benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	   30000	    219280 ns/op
BenchmarkCopyDir-64                      	      10	 813604712 ns/op


e29bba98e1 (HEAD) copy: Refactor, and separate out metadata copy
$ TMPDIR=/mnt/bench ./copy.e29bba98e1  -test.run none  -test.bench=.* -test.benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	    2000	   2523843 ns/op
BenchmarkCopyDir-64                      	       5	1503540458 ns/op
PASS
$ ./copy.e29bba98e1  -test.run none  -test.bench=.* -test.benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-64    	    3000	   1697842 ns/op
BenchmarkCopyDir-64                      	       5	1298545445 ns/op
PASS
```

dmcgowan commented:
LGTM

Thanks for keeping the interface the same, documenting the code, adding tests, and benchmarks

sargun commented:
@dmcgowan Mind nudging janky? Also, I'll freeze the PR as it is now, unless you want me to add more docs. If so, want to move it to review?

sargun commented:
I'm going to follow this up with the OpenAt patches.

It'll get rid of:
* Extra allocations due to filepath manipulation
* go calling epoll_ctl (this takes ~200ns per file)

thaJeztah reviewed (approved):
SGTM (thanks for reviewing @dmcgowan)

AkihiroSuda commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -11,20 +11,32 @@ package copy // import "github.com/docker/docker/daemon/graphdriver/copy"
+	// and we have to force-set them.
+	desiredUmask = 7022
+	// This is based on some empirical testing on EXT4 filesystems
+	workerCount = 8
```
Maybe this should be configurable via --storage-opt?

thaJeztah replied:
Do you foresee negative side-effects to this change? i.e.; is there a specific reason you want this to be configurable @AkihiroSuda ?

I'm not _against_ making things tuneable, but also want to guard against adding an option for _everything_ (also because work is in progress to delegate more things to containerd)

AkihiroSuda replied:
Wondering high concurrencycan be problematic for poor machines, but no benchmark data

AkihiroSuda replied:
Wondering high concurrencycan be problematic for poor machines, but no benchmark data

kolyshkin replied:
Perhaps `workerCount` should be capped to number of CPUs on the machine?

cpuguy83 replied:
Number of CPU's gives worse performance (unless of course nCPU > 8)... on a single core system is WAY worse
 
This is from the branch head compared to setting the worker count to the number of CPU's on a single core system
```
benchmark                              old ns/op     new ns/op     delta
BenchmarkCopyDirFewerDirsMoreFiles     340916        4474857       +1212.60%
```

This issue here is this is all I/O and not CPU bound and so the runtime is just waiting there are too few workers.

sargun replied:
So, at least on XFS, setting workers to more than 8 has degradation. Setting it to 64 (our CPU count) hurts pretty bad.

cpuguy83 replied:
In this case do you think it's worthwhile to add a storage-opt for vfs?

cpuguy83 replied:
Oh yeah and just a note, I ran these benches using ext4.

sargun replied:
It sounds like giving the user this knob would be counter productive, except for low memory machines. I would suggest making the channel depth, and worker count environment variables, if we really want to.

cpuguy83 replied:
Or a storage-opt? Could be nice for people with beefy machines as well.

As I recall, a single worker was *much* slower with this change as compared to the original.

sargun commented:
Can someone with tiny physical hardware, or spinning disks test this?

On Thu, Oct 25, 2018 at 8:33 AM Akihiro Suda <notifications@github.com>
wrote:

> *@AkihiroSuda* approved this pull request.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/moby/moby/pull/38034#pullrequestreview-168446984>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAtyREb0Y0CBFfCXAigehdsNcO0CorRrks5uodncgaJpZM4XahY0>
> .
>

cpuguy83 commented:
I'll fire this up on since and dual core systems with spinning disks.

sargun commented:
At least in the metadata heavy use case, you basically are serialized by
the file system's locking. In the file heavy case, you have a lot of tiny
locks, and are mostly doing I/O. In the I/O heavy case, you're I/O
contended. The concurrency here isn't to suck as much from the CPU cores as
possible, but it is to try to best take advantage of the locking behaviour.

On Thu, Oct 25, 2018 at 11:48 AM Kirill Kolyshkin <notifications@github.com>
wrote:

> *@kolyshkin* commented on this pull request.
> ------------------------------
>
> In daemon/graphdriver/copy/copy.go
> <https://github.com/moby/moby/pull/38034#discussion_r228292252>:
>
> >  	"golang.org/x/sys/unix"
>  )
>
> +const (
> +	// Docker sets this earlier on in initialization to 0022
> +	// Unfortunately, sometimes the special bits are not respected
> +	// and we have to force-set them.
> +	desiredUmask = 7022
> +	// This is based on some empirical testing on EXT4 filesystems
> +	workerCount = 8
>
> Perhaps workerCount should be capped to number of CPUs on the machine?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/moby/moby/pull/38034#discussion_r228292252>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAtyRKY_tvVWCoK6nAkjqj2FxzOKEZd0ks5uogdugaJpZM4XahY0>
> .
>

cpuguy83 commented:
dual core:

```
~/go/src/github.com/docker/docker/daemon/graphdriver/copy$ go test -bench Bench --benchtime 10s
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles-2   	   10000	   1186464 ns/op
BenchmarkCopyDir-2                     	      20	1119077321 ns/op
PASS
ok  	github.com/docker/docker/daemon/graphdriver/copy	442.613s
```

Single core... times out with a bench time of 10s.
Here it is with no explicit bench time:
```
~/go/src/github.com/docker/docker/daemon/graphdriver/copy$ go test -bench Bench
goos: linux
goarch: amd64
pkg: github.com/docker/docker/daemon/graphdriver/copy
BenchmarkCopyDirFewerDirsMoreFiles 	    3000	    344618 ns/op
BenchmarkCopyDir                   	       1	3101725607 ns/op
PASS
ok  	github.com/docker/docker/daemon/graphdriver/copy	115.835s
```

sargun commented:
@cpuguy83 Can you also run the benchmark on commit e29bba98e19767110bc858a3173e480fc243855c? I added the benchmarks there doing single core, and made a small optimization around metadata copies.

cpuguy83 commented:
10s bench time times out on dual core now, here is no explicit bench time set between the two commits:

```
benchmark                                old ns/op     new ns/op     delta
BenchmarkCopyDirFewerDirsMoreFiles-2     2195113       308630        -85.94%
```

Single core:
```
benchmark                              old ns/op     new ns/op     delta
BenchmarkCopyDirFewerDirsMoreFiles     280491        340916        +21.54%
```

sargun commented:
I'd be fine with introducing an environment variable, or global variable to limit workers. I do not want to use storage-opt because eventually I want to move this in the direction of having a shared worker pool for all of the copies.

thaJeztah commented:
> I'd be fine with introducing an environment variable, or global variable to limit workers. I do not want to use storage-opt because eventually I want to move this in the direction of having a shared worker pool for all of the copies.

@cpuguy83 @AkihiroSuda @dmcgowan wdyt? are you ok with an environment variable?

sargun commented:
I synced with @cpuguy83 offline. I'm going to introduce a storage opt `copy.concurrency`, and allow for it to be passed through VFS, and legacy overlay.

Commit:
layer, vfs: Remove switching copy method during tests

This removes a legacy change where the layer tests
temporarily changed the dirCopy mechanism for the
VFS driver.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

Commit:
vfs: Allow users to specify copy concurrency

This allows users to specify the copy concurrency in the VFS
graphdriver using copy.concurrency.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

Commit:
overlay: Allow specifying copy concurrency

This adds the driver opt to the overlay(1) driver
to allow specifying copy.concurrency.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

AkihiroSuda commented on daemon/graphdriver/vfs/driver.go:
```diff
@@ -30,6 +29,27 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
+				return nil, errors.Wrapf(err, "Cannot parse copy.concurrency")
+			}
+			if d.copyConcurrency < 0 {
+				return nil, fmt.Errorf("Copy concurrency '%d' is less than 0", d.copyConcurrency)
```
nit: `error` should start with a lower character

AkihiroSuda commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -107,156 +119,395 @@ func copyXattr(srcPath, dstPath, attr string) error {
+	errGroup, errGroupCtx := errgroup.WithContext(ctx)
+	workers := make([]chan *fileCopyInfo, workerCount)
+	for i := 0; i < len(workers); i++ {
+		c := make(chan *fileCopyInfo, 100)
```
100 should be a named `const`. (no need to make this `storage-opt`, probably 😄 )

sargun commented:
@AkihiroSuda I addressed your comments.

sargun commented:
Yet another unrelated failure with Experimental (TestServiceWithDefaultAddressPoolInit).

thaJeztah commented:
@Selansen was looking into that test (https://github.com/moby/moby/pull/37836), but looks like it stalled. @Selansen you still working on that one? It's failing quite frequently

I'll restart the "experimental" CI

selansen commented:
I tried to reproduced locally but it never happened to me. Let me check the log files.

kolyshkin commented:
The only thing I don't like here is all the dances around `system.Chtimes` and `system.LUtimesNano` -- it seems that whether it is a symlink or not, `LUtimesNano` can be used (unconditionally), resulting in simpler code. But it was that way before this PR and can be addressed separately.

LGTM.

kolyshkin commented:
Also, maybe it makes sense to implement the changes proposed here in https://github.com/containerd/continuity/tree/master/fs instead, and use that package from here? WDYT @dmcgowan @AkihiroSuda @cpuguy83 @thaJeztah

cpuguy83 commented on daemon/graphdriver/vfs/copy_linux.go:
```diff
@@ -2,6 +2,6 @@ package vfs // import "github.com/docker/docker/daemon/graphdriver/vfs"
 import "github.com/docker/docker/daemon/graphdriver/copy"
 
-func dirCopy(srcDir, dstDir string) error {
+func dirCopy(srcDir, dstDir string, concurrency int) error {
```
Not using this concurrency value.

sargun replied:
Derp, will fix

sargun commented:
@kolyshkin I'm not going to add it in there. @cpuguy83  and I synched offline about this. currently there's a bunch of Docker specific stuff in there, and that library isn't really a general purpose library yet.

sargun commented:
@cpuguy83 fixed.

cpuguy83 reviewed (approved):
Couple of nits but LGTM.

It may be prudent to look into replacing the `readDirNames` implementation here (and investigate where else this might be an issue).

sargun commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -107,156 +120,395 @@ func copyXattr(srcPath, dstPath, attr string) error {
+		return err
+	}
+	// Check if the errgroup has gotten canceled / shutdown
+	err = w.ctx.Err()
```
This is just there to keep the code as "clean" as possible, and to make the code as similar to the code in the stdlib. It could get passed in as well.

cpuguy83 replied:
We should pass in in general. I think it's fine to deviate from the stdblib interface since we do have a context to pass around.

Commit:
copy: Allow API consumers to specify concurrency

This extends the API and adds the function DirCopyWithConcurrency,
which allows the user to specify the copy concurrency for odd
shaped hardware.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

Commit:
copy: Remove the usage of sync.WaitGroup on every inode

We know based on stat from the source inode whether or not it is
hardlinked based on the nlinks field. Iff this is greater than 1
then we send over a shared inode with a mutex. This also reduces
the total amount of memory / book keeping that needs to be done
by the worker threads, and the logic to handle these hard links
is in the main walkr.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

sargun commented:
Can someone nudge CI? @AkihiroSuda

thaJeztah commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -107,156 +120,393 @@ func copyXattr(srcPath, dstPath, attr string) error {
+// concurrency level. Specifying workerCount as 0 will make it use the default worker
+// count
+func DirCopyWithConcurrency(srcDir, dstDir string, copyMode Mode, copyXattrs bool, workerCount int) error {
+	if workerCount < 0 {
```
Can we use an `uint`, if we don't accept negative values?

thaJeztah commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -107,156 +120,395 @@ func copyXattr(srcPath, dstPath, attr string) error {
+	if workerCount == 0 {
+		workerCount = defaultWorkerCount
+	}
+	ctx, cancel := context.WithCancel(context.Background())
```
Looks like this wasn't addressed yet

thaJeztah commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -107,156 +120,395 @@ func copyXattr(srcPath, dstPath, attr string) error {
+	if err != nil {
+		return nil, err
+	}
+	names, err := f.Readdirnames(-1)
```
@cpuguy83 this still a concern to you? (and what's your suggestion for an alternative?)

thaJeztah commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -107,156 +120,393 @@ func copyXattr(srcPath, dstPath, attr string) error {
-				return err
-			}
+// readDirNames reads the directory named by dirname and returns
+// a sorted list of directory entries.
```
Is this accurate? I don't see where it's sorted (nor does Golang's description mention sorting); https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/os/dir.go#L29-L40

Not a big issue, just double-checking so that we don't document behavior that isn't guaranteed / part of the contract

thaJeztah commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -107,156 +120,393 @@ func copyXattr(srcPath, dstPath, attr string) error {
+// DirCopyWithConcurrency performs the same work as DirCopy, but allows you to specify the
+// concurrency level. Specifying workerCount as 0 will make it use the default worker
+// count
+func DirCopyWithConcurrency(srcDir, dstDir string, copyMode Mode, copyXattrs bool, workerCount int) error {
```
Wondering why you're using different terminology here than in daemon/graphdriver/vfs/copy_linux.go `dirCopy()` (`workercount` vs `concurrency`). Perhaps "workercount" should be seen as an implementation detail, and this should just be named `concurrency` (after all, that's also part of the function name).

(open for input though 😅)

thaJeztah commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -107,156 +120,393 @@ func copyXattr(srcPath, dstPath, attr string) error {
+// DirCopy copies or hardlinks the contents of one directory to another,
+// properly handling xattrs, and soft links
+//
+// Copying xattrs can be opted out of by passing false for copyXattrs.
```
Hm.. (not new, so no need to change as part of this PR), but this wording is a bit odd; given that a `bool` defaults to `false`, this feels more like an "opt-in" 🤷‍♂️

thaJeztah commented on daemon/graphdriver/overlay/overlay.go:
```diff
@@ -181,23 +176,27 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
 		}
 		key = strings.ToLower(key)
 		switch key {
+		case "copy.concurrency":
```
This option won't be available to the `overlay2` storage driver, correct?

If so; wondering if we should add it to the `overlay` storage driver, given that we marked that driver deprecated.

thaJeztah commented on daemon/graphdriver/vfs/driver.go:
```diff
@@ -30,6 +29,27 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
+		}
+		key = strings.ToLower(key)
+		switch key {
+		case "copy.concurrency":
```
Wondering if this should this be `vfs.copy-concurrency` (if it's specific to the `vfs` driver) 🤔 @cpuguy83 ?

thaJeztah reviewed (changes requested):
left some questions inline 😅

this will also need changes to the documentation in https://github.com/docker/cli/blob/master/docs/reference/commandline/dockerd.md (to document the `copy.concurrency` configuration option

AkihiroSuda commented:
carried in https://github.com/moby/moby/pull/40521
